### PR TITLE
8296374: Check for young region in G1BarrierSet::invalidate instead of card-by-card check

### PR DESCRIPTION
**Pull request by walulyai**
Hi,

Please review this change to young region card checks in G1BarrierSet::invalidate(). A single check should suffice because the MemRegion for invalidation in G1 is either completely in the young gen or not. 

Testing: Tier 1-3

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296374](https://bugs.openjdk.org/browse/JDK-8296374): Check for young region in G1BarrierSet::invalidate instead of card-by-card check

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [58bf53a8](https://git.openjdk.org/jdk/pull/11677/files/58bf53a893d6f3a228e33d34100a0b98459a5b07)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11677/head:pull/11677` \
`$ git checkout pull/11677`

Update a local copy of the PR: \
`$ git checkout pull/11677` \
`$ git pull https://git.openjdk.org/jdk pull/11677/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11677`

View PR using the GUI difftool: \
`$ git pr show -t 11677`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11677.diff">https://git.openjdk.org/jdk/pull/11677.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1BarrierSet.cpp [hunk 1]
```diff
@@ -107,8 +107,12 @@ void G1BarrierSet::invalidate(MemRegion mr) {
   }
   volatile CardValue* byte = _card_table->byte_for(mr.start());
   CardValue* last_byte = _card_table->byte_for(mr.last());
-  // skip initial young cards
-  for (; byte <= last_byte && *byte == G1CardTable::g1_young_card_val(); byte++);
+
+  // skip young cards
+  if (*byte == G1CardTable::g1_young_card_val()) {
+    guarantee(*last_byte == G1CardTable::g1_young_card_val(), "MemRegion should be either completely in young gen or not");
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1BarrierSet.cpp:113]: I think a stronger assertion is that `mr` is inside a heap-region, sth along the line of `mr.size() < HeapRegionSize`.

File: src/hotspot/share/gc/g1/g1BarrierSet.cpp [hunk 2]
```diff
@@ -107,8 +107,12 @@ void G1BarrierSet::invalidate(MemRegion mr) {
...
+    return;
+  }
 
   if (byte <= last_byte) {
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1BarrierSet.cpp:117]: This check can be removed now, right?

File: src/hotspot/share/gc/g1/g1BarrierSet.cpp [hunk 3]
```diff
@@ -107,22 +107,24 @@ void G1BarrierSet::invalidate(MemRegion mr) {
...
-    for (; byte <= last_byte; byte++) {
-      CardValue bv = *byte;
-      if ((bv != G1CardTable::g1_young_card_val()) &&
-          (bv != G1CardTable::dirty_card_val())) {
-        *byte = G1CardTable::dirty_card_val();
-        qset.enqueue(queue, byte);
-      }
+  // skip young gen cards
+  if (*byte == G1CardTable::g1_young_card_val()) {
+    assert(mr.word_size() <= HeapRegion::GrainWords, "Invalidated MemRegion is larger than HeapRegion in young gen");
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BarrierSet.cpp:113]: I would change this to a check that the memregion (in young) does not span multiple regions; for the same reason one could check that it is not humongous(is that true?), but I'm not sure what the background for this check is, so.
Feel free to improve the check.
Other than that comment, looks good.

**Review by tschatzl — APPROVED** [review]

**Review by tschatzl — APPROVED** [review]
Better :)

**Comment by walulyai** [comment]
Thanks for the reviews.

/integrate

**Pull request closed without merging**